### PR TITLE
Idea new `state` and `meteadata` feature

### PR DESCRIPTION
**Pull request by Almazatun**
Resolves # .

- Idea new `state` and `metadata` feature #1135
- Update idea `e2e` tests

@reviewer-or-team

### CODE REVIEW

File: idea/data-storage/src/database/entities/meta.entity.ts [hunk 1]
```diff
@@ -7,23 +7,23 @@ import { Program } from './program.entity';
...
+  public owner: string;
 
   @Column({ nullable: true })
-    meta: string;
+  public hex: string;
 
-  @Column({ nullable: true })
-    metaWasm: string;
+  @Column({ nullable: true, type: 'json' })
+  public data: string;
```
Comment (osipov-mit) [inline idea/data-storage/src/database/entities/meta.entity.ts:22]: I'm not sure that `data` is the best option.
Maybe `types` or smth like this?

File: idea/data-storage/src/database/entities/state.entity.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  public name: string;
+
+  @Column()
+  public wasmBuffBase64: string;
+
+  @Column({ type: 'json' })
+  public funcNames: string;
+
+  @Column({ type: 'json' })
+  public functions: string;
```
Comment (osipov-mit) [inline idea/data-storage/src/database/entities/state.entity.ts:21]: What is the diferrence between these fields?

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: idea/data-storage/src/database/entities/state.entity.ts (hunk 2 above)
Comment (Almazatun) [reply idea/data-storage/src/database/entities/state.entity.ts:21]: Include this field for making able to do query requests with a `function name`. 

Before including this field I tested it, it is possible to make a query request by `function` name when in our field `functions` will have a structure like the below.

```javascript
State { functions : { func_1: some_value, func_2: some_value, func_3: some_value }}
```

Query request didn't work if we want to get by a key of the `object` but by `value` of key, it works.
Example: 

 #### Search by key in JSON field
  🔴 Didn't work  
```javascript
 Object { [search_by_key]: some_value } 
 ``` 
 #### Search by value in JSON field
 ✅ Works  
  ``` javascript 
  Object { key: search_by_value } or { key: [func_1, func_2, func_3] }
  ```

File: idea/data-storage/src/database/entities/meta.entity.ts (hunk 1 above)
Comment (Almazatun) [reply idea/data-storage/src/database/entities/meta.entity.ts:22]: Alright, `types` sounds better.

File: idea/data-storage/src/gear/gear-event-listener.ts [hunk 3]
```diff
@@ -65,6 +66,11 @@ export class GearEventListener {
     }
   }
 
+  public async validateMetaHex(hex: string, programId: string): Promise<boolean> {
+    const metaHash = await this.api.program.metaHash(programId as Hex);
+    return metaHash === blake2AsHex(hex, 256);
+  }
```
[the remark this answers is not in the document]
Comment (Almazatun) [reply idea/data-storage/src/gear/gear-event-listener.ts:72]: Will change to `flag` function

File: idea/data-storage/src/database/entities/meta.entity.ts [hunk 4]
```diff
@@ -7,23 +7,23 @@ import { Program } from './program.entity';
 @Entity()
 export class Meta implements IMeta {
   @PrimaryGeneratedColumn('rowid')
-    id: string;
+  public  id: string;
 
   @Column()
-    program: string;
+  public program: string;
```
[the remark this answers is not in the document]
Comment (Almazatun) [reply idea/data-storage/src/database/entities/meta.entity.ts]: Alright

File: idea/common/src/enums/api-methods.enum.ts [hunk 5]
```diff
@@ -1,6 +1,9 @@
 enum API_METHODS {
   PROGRAM_DATA = 'program.data',
   PROGRAM_META_ADD = 'program.meta.add',
+  PROGRAM_STATE_ADD = 'program.state.add',
```
Comment (osipov-mit) [inline idea/common/src/enums/api-methods.enum.ts:4]: We should be able to get metadata and state wasms for particular code too.

File: idea/common/src/interfaces/meta.ts [hunk 6]
```diff
@@ -1,7 +1,19 @@
...
+
+export interface IMetaData {
+  init: HumanTypesRepr;
+  handle: HumanTypesRepr;
+  reply: HumanTypesRepr;
+  others: HumanTypesRepr;
+  signal: number | null;
+  state: number | null;
+}
+
```
Comment (osipov-mit) [inline idea/common/src/interfaces/meta.ts:14]: You can import these interfaces from @gear-js/api

File: idea/data-storage/src/common/helpers/get-hex-wasm-state.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { CreateType, Hex } from '@gear-js/api';
+import * as dotenv from 'dotenv';
+
+dotenv.config();
+
+export function getHexWasmState(wasmStateBuff: Buffer): Hex {
+  if(process.env.TEST_ENV_UNIT) return 'hex' as Hex;
+
+  return CreateType.create('Bytes', wasmStateBuff).toHex();
+}
```
Comment (osipov-mit) [inline idea/data-storage/src/common/helpers/get-hex-wasm-state.ts:10]: It much better to compare wasms by their's hash, not full wasm in hex representaion. For this purpose you can use `generateCodeHash` function from @gear-js/api

File: idea/data-storage/src/common/helpers/get-state-meta.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import { getStateMetadata } from '@gear-js/api';
+import { config } from 'dotenv';
+
+config();
+
+export async function getStateMeta(stateMetaBuff: Buffer): Promise<{ functions: object }> {
```
Comment (osipov-mit) [inline idea/data-storage/src/common/helpers/get-state-meta.ts:6]:
```suggestion
export async function getStateMeta(stateMetaBuff: Buffer): Promise<StateMetadata> {
```

File: idea/data-storage/src/database/entities/code.entity.ts [hunk 9]
```diff
@@ -5,32 +5,39 @@ import { BaseEntity } from './base.entity';
...
   @Column({ type: 'enum', enum: CodeStatus })
-    status: CodeStatus;
+  public status: CodeStatus;
 
   @Column({ nullable: true })
-    expiration: string;
+  public expiration: string;
+
+  @Column({ nullable: true })
+  public hex: string;
```
Comment (osipov-mit) [inline idea/data-storage/src/database/entities/code.entity.ts:32]: Is this field necessary?

File: idea/data-storage/src/database/entities/meta.entity.ts [hunk 10]
```diff
@@ -7,23 +7,17 @@ import { Program } from './program.entity';
...
-  @Column()
-    program: string;
-
-  @Column()
-    owner: string;
+  public  id: string;
 
   @Column({ nullable: true })
-    meta: string;
+  public hex: string;
```
Comment (osipov-mit) [inline idea/data-storage/src/database/entities/meta.entity.ts:13]: Let's call it hash.

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: idea/common/src/enums/api-methods.enum.ts (hunk 5 above)
Comment (Almazatun) [reply idea/common/src/enums/api-methods.enum.ts:4]: Done

File: idea/common/src/interfaces/meta.ts (hunk 6 above)
Comment (Almazatun) [reply idea/common/src/interfaces/meta.ts]: Done

File: idea/data-storage/src/common/helpers/get-state-meta.ts (hunk 8 above)
Comment (Almazatun) [reply idea/data-storage/src/common/helpers/get-state-meta.ts]: Done

File: idea/data-storage/src/database/entities/meta.entity.ts (hunk 10 above)
Comment (Almazatun) [reply idea/data-storage/src/database/entities/meta.entity.ts]: Done

**Review by osipov-mit — APPROVED** [review]

**Pull request merged**